### DISCUSSION TITLE
x-man-page:// URL scheme not picking up man pages from brew?

### DISCUSSION THREAD

**Discussion by gingerbeardman** [Homebrew/discussions#3425] [Everyday usage]
Looking for some sanity check on this...

## URL scheme
- `x-man-page://gifsicle`
- Result: "No manual entry for gifsicle"
- [Screenshot](https://user-images.githubusercontent.com/49612/178609787-3e4af37f-1246-446d-ade2-4b6e69781f50.png)

## Terminal
- `man gifsicle`
- Result: man page displayed OK
- [Screenshot](https://user-images.githubusercontent.com/49612/178609974-40f8bbca-a7c2-4e1f-acc7-33d98a31ab1a.png)

## Location
- `/opt/homebrew/share/man/man1/gifsicle.1`

## Setup
- I'm on macOS 12.4 (21F79) Monterey
- zsh shell

**Comment by Bo98** [comment]
It probably doesn't pick up `MANPATH` from `brew shellenv` in your profile so you may need to modify `/etc/man.conf` instead.

**Reply by gingerbeardman** [reply]
Hmm, interesting.

My MANPATH is currently:

```
$ manpath
/opt/homebrew/share/man:/usr/local/share/man:/opt/local/share/man:/usr/share/man:/opt/X11/share/man:/Library/Apple/usr/share/man:/Library/Frameworks/Mono.framework/Versions/Current/share/man:/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/share/man:/Applications/Xcode.app/Contents/Developer/usr/share/man:/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/share/man
```

and `brew shellenv` gives:

```
$ brew shellenv
export HOMEBREW_PREFIX="/opt/homebrew";
export HOMEBREW_CELLAR="/opt/homebrew/Cellar";
export HOMEBREW_REPOSITORY="/opt/homebrew";
export PATH="/opt/homebrew/bin:/opt/homebrew/sbin${PATH+:$PATH}";
export MANPATH="/opt/homebrew/share/man${MANPATH+:$MANPATH}:";
export INFOPATH="/opt/homebrew/share/info:${INFOPATH:-}";
```

So it seems the brew location is already in my MANPATH? 

Let me know if I misunderstand this. I've never heard of MANPATH before today :)

**Reply by gromgit** [reply]
Assuming you tried to `open x-man-page://gifsicle`, the reason for the behavior you're seeing is a bit convoluted:
* `open` (IIRC) uses [Launch Services](https://developer.apple.com/documentation/coreservices/launch_services)
* ...which commands `launchservicesd`, which is a root-owned daemon started on boot rather than your login (so it has a completely separate environment from your existing nteractive Terminal session)
* ...which launches a new Terminal session under your UID
* ...which starts a _non-interactive shell session_ (so none of your shell startup files are sourced)
* ...which runs `man gifsicle`, in a barebones environment that has no knowledge of Homebrew paths

So yeah, as @Bo98 said, you'd have to enhance `/etc/man.conf`, for what you want to happen.

_Sidenote:_ That part about "none of your shell startup files are sourced" is true if your login shell is Bash. If it's Zsh, you can try adding `eval "$(/opt/homebrew/bin/brew shellenv)"` to `${HOME}/.zshenv`--it normally doesn't exist. This file is sourced by all Zsh shell sessions, so this might want. The downside: this file is sourced by _all_ Zsh shell sessions, so putting too much stuff in here will make all your Zsh scripts start slower.

**Reply by gingerbeardman** [reply]
Correct, `open x-man-page://gifsicle` gives "No manual entry..."

Thanks for the hints, the issue is now resolved by doing this:

1. run `sudo pico /etc/man.conf`
2. add `MANPATH /opt/homebrew/share/man` under the other `MANPATH` entries
3. save and quit

Cheers!

**Reply by carlocab** [reply]
Note that Apple likes to roll back changes to `/etc/man.conf` on OS updates. I haven't worked out a way to make it stick.

**Reply by gingerbeardman** [reply]
Oh! Thanks for the headsup @carlocab. 

I'll keep that in mind. At least it's an easy/quick change to redo.

**Reply by carlocab** [reply]
One alternative to customising `/etc/man.conf` is to create a file `/etc/manpaths.d/brew` containing only
```
/opt/homebrew/share/man
```
That one should stick, I think. Even Apple isn't rude enough to delete the contents of `/etc/*.d/` directories on OS upgrades.

**Reply by gingerbeardman** [reply]
That one doesn't work for me.

**Reply by gromgit** [reply]
Yeah, unfortunately adding `/etc/manpaths.d/brew` only works if `path_helper` is invoked during shell startup...and since it's normally called in `/etc/profile` and `/etc/zprofile`, that will only happen for interactive shell sessions.

Figuring out what Bash and Zsh do is just "advanced science"...which looks an awful lot like "magic" if you're not in the mood. :grinning:

**Reply by gingerbeardman** [reply]
> Note that Apple likes to roll back changes to `/etc/man.conf` on OS updates. I haven't worked out a way to make it stick.

Quite correct, had to do this fix again after the recent macOS update.

**Reply by carlocab** [reply]
> Yeah, unfortunately adding `/etc/manpaths.d/brew` only works if `path_helper` is invoked during shell startup...and since it's normally called in `/etc/profile` and `/etc/zprofile`, that will only happen for interactive shell sessions.

Oops. You can always add a call to `/usr/libexec/path_helper` to `/etc/zshrc`. But I guess that goes back to the problem of Apple wiping files in `/etc` on OS upgrades.

**Reply by gingerbeardman** [reply]
A one line fix for this is `echo "MANPATH /opt/homebrew/share/man" | sudo tee -a /etc/man.conf` as it doesn't matter where the config line is in the file.
